### DISCUSSION TITLE
X coordinate negated when assembling the bottom

### DISCUSSION THREAD

**Discussion by jstanley0** [matthewlai/JLCKicadTools#55] [General]
I recently used this tool to order a PCB that was assembled on the top and it worked beautifully; however, I have another design where I'm trying to assemble the bottom of the board instead, and it always previews blank on JLCPCB's website (that is, no parts appear and I see an unpopulated circuit board). I compared the all-pos.csv file generated by KiCad to the cpl_jlc.csv file generated by this program and found the X coordinates were all negated. A glance through the code shows this is done intentionally. But when I _undid_ it, my board previewed correctly. (I am not quite ready to actually order this board yet so I can't say whether it makes it through JLC's DFM check.)

I'm gonna guess that at some point JLCPCB "fixed" this problem on their end, making the negation on our side unnecessary?

**Comment by matthewlai** [comment]
Thanks for the report. This is actually due to changes in how KiCad exports positions.

We added the negation in this commit: https://github.com/matthewlai/JLCKicadTools/commit/4d87eabb2ffcd5d163b3dd8b34f414a145b41351

Because KiCad was generating negative x coordinates, so the negation fixes that.

And KiCad did that because someone said they want their position files that way: https://forum.kicad.info/t/footprint-position-file-x-coordinates-suddenly-negative/23253

Now it looks like with 5.1.7, a checkbox has been added to set the behaviour: https://gitlab.com/kicad/code/kicad/-/issues/4638

I will try to implement automatic detection so we support both old and new KiCad versions.

**Reply by jstanley0** [reply]
Ah, somehow I missed that checkbox. But if I check "Use negative X coordinates for footprints on the bottom layer", the script works as expected. I think just adding a note to step 3 of the instructions on your blog would be sufficient.

Thank you!

**Reply by matthewlai** [reply]
All fixed now if you update :).
https://github.com/matthewlai/JLCKicadTools/pull/56

It should work automatically with both checked and unchecked.

**Reply by matthewlai** [reply]
Would appreciate if you can help test!

If you run pip install again it should pick up the new version.

**Reply by jstanley0** [reply]
I tried it out. Parts previewed in the correct places either way; however, with "Use negative X coordinates" unchecked, U1 (a TQFP-32 package) ended up rotated 180 degrees from where it should have been. A couple SOT-23 transistors looked fine either way. Nothing else in my circuit but resistors and ceramic capacitors so I didn't notice if they'd been rotated.
